src/javaRosa/util.js

```diff
@@ -0,0 +1,277 @@
+        if (!mug.options.canOutputValue) {
+            // TODO display message near where it was dropped
+            // HACK should be in the itext widget, which has the mug and path
+            var current = form.vellum.getCurrentlySelectedMug(),
+                typeName = mug.options.typeName;
+            if (current) {
+                current.addMessage(null, {
+                    key: "javaRosa-output-value-type-error",
+                    level: mug.WARNING,
+                    message: typeName + " nodes cannot be used in an " +
+                        "output value. Please remove the output value for " +
+                        "'" + path + "' or your form will have errors."
+                });
+            }
+        }
+    };
+
+    var getOutputRef = function(path, dateFormat) {
+        if (dateFormat) {
+            return '<output value="format-date(date(' + path + '), \'' + dateFormat + '\')"/>';
+        } else {
+            return '<output value="' + path + '" />';
+        }
+    };
+
+    var warnOnCircularReference = function(property, mug, path, refName, propName) {
+        // TODO track output refs in logic manager
+        if (path === "." && property === 'label') {
+            var fieldName = mug.p.getDefinition(property).lstring;
+            mug.addMessage(propName, {
+                key: "core-circular-reference-warning",
+                level: mug.WARNING,
+                message: "The " + fieldName + " for a question " +
+                    "is not allowed to reference the question itself. " +
+                    "Please remove the " + refName + " from the " +
+                    fieldName +" or your form will have errors."
+            });
+        }
+    };
+
+    var insertOutputRef = function(vellum, target, path, mug, dateFormat) {
+        var output = getOutputRef(path, dateFormat),
+            form = vellum.data.core.form;
+        if ((!mug && vellum.opts().features.rich_text) || (mug && mug.supportsRichText())) {
+            richText.editor(target).insertOutput(output);
+        } else {
+            util.insertTextAtCursor(target, output, true);
+        }
+        if (mug) {
+            warnOnCircularReference('label', mug, path, 'output value', target.attr('name'));
+            warnOnNonOutputableValue(form, mug, path);
+        }
+    };
+
+    return {
+        ITEXT_PROPERTIES: ITEXT_PROPERTIES,
+        SUPPORTED_MEDIA_TYPES: SUPPORTED_MEDIA_TYPES,
+        forEachItextItem: forEachItextItem,
+        generateItextXLS: generateItextXLS,
+        getDefaultItextRoot: getDefaultItextRoot,
+        getDefaultItextId: getDefaultItextId,
+        getItextItemsFromMugs: getItextItemsFromMugs,
+        getOutputRef: getOutputRef,
+        insertOutputRef: insertOutputRef,
+        looksLikeMarkdown: looksLikeMarkdown,
+        parseXLSItext: parseXLSItext,
+        warnOnCircularReference: warnOnCircularReference,
+        warnOnNonOutputableValue: warnOnNonOutputableValue,
```

Comment:
`warnOnCircularReference` and `warnOnNonOutputableValue` appear to be only referenced by `insertOutputRef`, so they don't need to be exported in the module object.